nibabel/brikhead.py

```diff
@@ -0,0 +1,487 @@
+
+    Examples
+    --------
+    >>> brik = load('example4d+orig.BRIK.gz')
+    >>> brik.shape
+    (33, 41, 25, 3)
+    >>> brik.affine
+    array([[ -3.    ,  -0.    ,  -0.    ,  49.5   ],
+           [ -0.    ,  -3.    ,  -0.    ,  82.312 ],
+           [  0.    ,   0.    ,   3.    , -52.3511],
+           [  0.    ,   0.    ,   0.    ,   1.    ]])
+    >>> head = load('example4d+orig.HEAD')
+    >>> np.array_equal(head.get_data(), brik.get_data())
+    True
+    """
+
+    header_class = AFNIHeader
+    valid_exts = ('.brik', '.head')
+    files_types = (('image', '.brik'), ('header', '.head'))
+    _compressed_suffixes = ('.gz', '.bz2')
+    makeable = False
+    rw = False
+    ImageArrayProxy = AFNIArrayProxy
+
+    @classmethod
+    @kw_only_meth(1)
+    def from_file_map(klass, file_map, mmap=True):
+        """
+        Parameters
+        ----------
+        file_map : dict
+            dict with keys ``image, header`` and values being fileholder
+            objects for the respective BRIK and HEAD files
+        mmap : {True, False, 'c', 'r'}, optional, keyword only
+            `mmap` controls the use of numpy memory mapping for reading image
+            array data.  If False, do not try numpy ``memmap`` for data array.
+            If one of {'c', 'r'}, try numpy memmap with ``mode=mmap``.  A
+            `mmap` value of True gives the same behavior as ``mmap='c'``.  If
+            image data file cannot be memory-mapped, ignore `mmap` value and
+            read array from file.
+        """
+        with file_map['header'].get_prepare_fileobj('rt') as hdr_fobj:
+            hdr = klass.header_class.from_fileobj(
+                hdr_fobj)
+        brik_fobj = file_map['image'].get_prepare_fileobj()
+        data = klass.ImageArrayProxy(brik_fobj, hdr,
+                                     mmap=mmap)
+        return klass(data, hdr.get_affine(), header=hdr, extra=None,
+                     file_map=file_map)
+
+    @classmethod
+    @kw_only_meth(1)
+    def from_filename(klass, filename, mmap=True):
+        """
+        Parameters
+        ----------
+        file_map : dict
+            dict with keys ``image, header`` and values being fileholder
+            objects for the respective BRIK and HEAD files
+        mmap : {True, False, 'c', 'r'}, optional, keyword only
+            `mmap` controls the use of numpy memory mapping for reading image
+            array data.  If False, do not try numpy ``memmap`` for data array.
+            If one of {'c', 'r'}, try numpy memmap with ``mode=mmap``.  A
+            `mmap` value of True gives the same behavior as ``mmap='c'``.  If
+            image data file cannot be memory-mapped, ignore `mmap` value and
+            read array from file.
+        """
+        file_map = klass.filespec_to_file_map(filename)
+        # only BRIK can be compressed, but `filespec_to_file_map` doesn't
+        # handle that case; remove potential compression suffixes from HEAD
```

Comment:
As a point of personal preference, I would rather improve `filespec_to_filemap` to handle the case where header may not be compressed. (But if that makes more sense as a separate PR, just to get this in, that's fine by me.)

Reply:
I think it'd be good to see this PR merged, first, if that's all right! Once this is in I'll be happy to raise an issue to discuss this change, since, as far as I can tell `filespec_to_file_map` is used by most image classes in nibabel.

Reply:
How about a little wrapper around filespec_to_filemap in this PR, to point the way, and keep that part of the logic separate?